### PR TITLE
Tuesday Changes, notably Proposal refactor

### PR DESCRIPTION
**Pull request by tbrent**
Collection of changes this PR should reflect:

- Refactor `Proposal` to Factory design in order to reduce `Manager` bytecode
- Use `trusted` prefix in all places where we are trusting another contract
- Simplify `_executeBasket` by passing in only `uint256` instead of `Basket`
- Fix typo in `Manager._weighted`
- Break `Manager` pausing into full-contract pausing and issuance-specific pausing
- Fix `isFullyCollateralized` unit comment
- Use SafeMath for time calculations
- Wrap `isFullyCollateralized` as a modifier and add to all core `Manager` functions
- Make `Manager.operator` public
- Emit events for `Vault.withdrawTo`
- Emit events for `Proposal`s
- Add and use `paused` modifier to mirror `notPaused`
- Require proposals to be non-empty
- Add error string to require in `Vault.changeManager`
- Set operator in the `Manager` constructor
- Make `ReserveEternalStorage` also inherit from `Ownable`
- Make `Ownable`'s `_owner` private and add getter
- Require baskets to always be of size 100 or less
- Unify `scaleFactor`, `divisor`, and `decimalsDivisor`
- Make `Makefile` _much_ more robust and deprecate old `.json` generation method
- Add documentation to clarify how `SwapProposal` relates to RSV total supply
- Change how we assert events to support both "there exists these events" and "the events are precisely these"
- Additions of tests where necessary
- Changes to indentation to keep maxchars under 100 and maintain readability

### CODE REVIEW

File: contracts/Proposal.sol [hunk 1]
```diff
@@ -25,7 +25,47 @@ import "./Basket.sol";
...
+    function createSwapProposal(address,
+        address[] calldata tokens,
+        uint256[] calldata amounts,
+        bool[] calldata toVault
+    ) external returns (IProposal);
+
+    function createWeightProposal(address proposer, Basket basket) external returns (IProposal);
+}
+
+contract ProposalFactory {
```
Comment (fiddlemath) [inline contracts/Proposal.sol]: ProposalFactory methods can be called by anyone. Is this fine? (Probably, but double-check.)

File: contracts/Proposal.sol [hunk 2]
```diff
@@ -111,36 +158,34 @@ contract SwapProposal is Proposal {
...
                 bool[] memory _toVault )
         Proposal(_proposer) public
     {
+        require(_tokens.length > 0, "proposal cannot be empty");
         require(_tokens.length == _amounts.length && _amounts.length == _toVault.length,
                 "unequal array lengths");
-        tokens = _tokens;
-        amounts = _amounts;
-        toVault = _toVault;
+        (tokens, amounts, toVault) = (_tokens, _amounts, _toVault);
```
Comment (fiddlemath) [inline contracts/Proposal.sol]: We should no longer need to compress this for Bytecode, and three lines is easier to inspect visually.

File: contracts/Proposal.sol [hunk 3]
```diff
@@ -111,36 +158,34 @@ contract SwapProposal is Proposal {
...
     }
 
     /// Return the newly-proposed basket, based on the current vault and the old basket.
-    function _newBasket(IRSV rsv, Basket oldBasket) internal returns(Basket) {
+    function _newBasket(IRSV rsv, Basket trustedOldBasket) internal returns(Basket) {
```
Comment (fiddlemath) [inline contracts/Proposal.sol]: Aw yeah, Good Old Basket.

File: contracts/rsv/Reserve.sol [hunk 4]
```diff
@@ -93,7 +94,7 @@ contract Reserve is IERC20, Ownable {
 
     /// Modifies a function to only run if sent by `role` or the contract's `owner`.
     modifier onlyOwnerOr(address role) {
-        require(msg.sender == _owner || msg.sender == role, "unauthorized: not role holder and not owner");
+        require(msg.sender == owner() || msg.sender == role, "unauthorized: not role holder and not owner");
```
Comment (fiddlemath) [inline contracts/rsv/Reserve.sol]: Could shorten this error message to "unauthorized: not owner or role". Clear _enough_, and <= 32 chars.

File: tests/base.go [hunk 5]
```diff
@@ -92,21 +94,41 @@ func (s *TestSuite) requireTxStrongly(tx *types.Transaction, err error) func(ass
...
+// error is nil. This signature allows the function to directly wrap our
+// abigen'd mutator calls.
+//
+// requireTxWithEvents returns a closure that can be used to assert the list of events
+// that were emitted during the transaction. This API is a bit weird -- it would
+// be more natural to pass the events in to the `requireTxWithEvents` call itself -- but
+// this is the cleanest way that is compatible with directly wrapping the abigen'd
+// calls, without using intermediate placeholder variables in calling code.
+//
+// Note: This closure asserts that the found events contain all of the expected events.
```
Comment (fiddlemath) [inline tests/base.go]: Update name of function in these comments.

File: tests/base.go [hunk 6]
```diff
@@ -62,17 +62,19 @@ type TestSuite struct {
 
 var coverageEnabled = os.Getenv("COVERAGE_ENABLED") != ""
 
-// requireTxStrongly requires that a transaction is successfully mined and does
+// requireTxWithEvents requires that a transaction is successfully mined and does
 // not revert. It also takes an extra error argument, and checks that the
 // error is nil. This signature allows the function to directly wrap our
 // abigen'd mutator calls.
```
Comment (fiddlemath) [inline tests/base.go]: This can be stated more clearly. I'll try...

File: tests/base.go [hunk 7]
```diff
@@ -157,15 +179,7 @@ func (s *TestSuite) assertManagerCollateralized() {
 }
 
 // assertCurrentBasketMirrorsTargets asserts that the current manager basket matches expectations.
```
Comment (fiddlemath) [inline tests/base.go]:
```suggestion
// assertBasket asserts that the current Manager's basket satisfies expectations.
```

File: tests/base.go [hunk 8]
```diff
@@ -368,15 +382,17 @@ func burningTransfer(from common.Address, value *big.Int) abi.ReserveTransfer {
 	}
 }
```
Comment (fiddlemath) [inline tests/base.go:386]:
```suggestion
// shiftRight returns `n`, shifted left by `decimals` zeroes.
```

File: tests/base.go [hunk 9]
```diff
@@ -368,15 +382,17 @@ func burningTransfer(from common.Address, value *big.Int) abi.ReserveTransfer {
...
 
-func toAtto(n uint32, decimals uint32) *big.Int {
+func shiftRight(n uint32, decimals uint32) *big.Int {
```
Comment (fiddlemath) [inline tests/base.go]: Should this not be "shiftLeft"?

File: contracts/Proposal.sol (hunk 1 above)
Comment (fiddlemath) [reply contracts/Proposal.sol]: Yeah, this is fine. Proposals can float around at random and never get into the RSV system. :+1:

File: contracts/Basket.sol [hunk 10]
```diff
@@ -46,6 +45,7 @@ contract Basket {
 
         // If there's a previous basket, copy those of its contents not already set.
         if (prev != Basket(0)) {
+            require(_tokens.length + prev.size() <= 100, "Basket: bad length");
```
Comment (fiddlemath) [inline contracts/Basket.sol]: If _tokens and prev.tokens contain some of the same addresses -- and you'd expect them to in most cases -- then this limit prevents them from constructing a basket of an OK length.

Comment (fiddlemath) [reply contracts/Basket.sol]: Recommend: insert this require, instead, at the end of this constructor. A list of the right length is constructed by then.

File: contracts/Manager.sol [hunk 11]
```diff
@@ -91,88 +93,119 @@ contract Manager is Ownable {
...
+    }
+
+    function _onlyOperator() internal view {
+        require(_msgSender() == operator, "operator only");
+    }
+
+    function _vaultCollateralized() internal view {
+        require(isFullyCollateralized(), "undercollateralized");
+    }
+
```
Comment (fiddlemath) [inline contracts/Manager.sol]: With the much smaller Manager bytecode that we now have, these can again be shorter, simpler modifiers.

File: contracts/Proposal.sol (hunk 3 above)
Comment (fiddlemath) [inline contracts/Proposal.sol]: `rsv` => `trustedRSV`

File: contracts/rsv/Reserve.sol [hunk 12]
```diff
@@ -27,7 +27,7 @@ contract Reserve is IERC20, Ownable {
 
 
     // Non-constant-sized data
-    ReserveEternalStorage internal data;
+    ReserveEternalStorage internal trustedData;
 
     // TX Fee helper contract
     ITXFee public txFee;
```
Comment (fiddlemath) [inline contracts/rsv/Reserve.sol]: Should be `trusted`

File: contracts/Basket.sol [hunk 13]
```diff
@@ -34,7 +34,6 @@ contract Basket {
     /// which to update `prev`. If `prev == address(0)`, act like it's an empty basket.
     constructor(Basket prev, address[] memory _tokens, uint256[] memory _weights) public {
```
Comment (fiddlemath) [inline contracts/Basket.sol]: `prev` => `trustedPrev`

**Review by fiddlemath — CHANGES_REQUESTED** [review]

File: tests/base.go (hunk 9 above)
Comment (tbrent) [reply tests/base.go]: I don't see this function in `tests/base.go` and that confuses me.

File: contracts/rsv/Reserve.sol (hunk 12 above)
Comment (tbrent) [reply contracts/rsv/Reserve.sol]: It is. I'm confused how you're seeing different files than I am

**Comment by tbrent** [comment]
Changes implemented. Not sure what was going on with our mismatching versions of the code. 

Edit: Ok, this was never happening, I'm crazy.

File: tests/base.go (hunk 9 above)
Comment (fiddlemath) [reply tests/base.go]: Er -- you don't see `shiftRight`?

File: contracts/rsv/Reserve.sol (hunk 12 above)
Comment (fiddlemath) [reply contracts/rsv/Reserve.sol]: Huh. I mean `txFee` should be `trusted`. `trustedData` already is `trusted`, yes.

Comment (fiddlemath) [reply contracts/rsv/Reserve.sol]: I probably misclicked the line, or maybe github moved the label around in response to some patch or another.

Comment (tbrent) [reply contracts/rsv/Reserve.sol]: Oh. No I'm not sure you did, I think I'm just a fool. 

Yes, `txFee` should be trusted. Pushing a fix now

File: tests/proposal_test.go [hunk 14]
```diff
@@ -34,12 +36,13 @@ type SwapProposalSuite struct {
...
 	proposalAddress common.Address
 	tokens          []common.Address
+	weights         []*big.Int
 	amounts         []*big.Int
 	toVault         []bool
 }
 
 var (
-	// Compile-time check that WeightProposalSuite implements the interfaces we think it does.
+	// Compile-futureTimecheck that WeightProposalSuite implements the interfaces we think it does.
```
Comment (fiddlemath) [inline tests/proposal_test.go]:
```suggestion
	// Compile-time check that WeightProposalSuite implements the interfaces we think it does.
```

File: tests/proposal_test.go [hunk 15]
```diff
@@ -111,27 +109,55 @@ func (s *SwapProposalSuite) TearDownSuite() {
...
+		},
 	)
 
 	// Check that proposer was set.
 	proposer, err := proposal.Proposer(nil)
 	s.Require().NoError(err)
 	s.Equal(s.proposer.address(), proposer)
 
-	// Check that time was not set.
+	// Check that futureTimewas not set.
```
Comment (fiddlemath) [inline tests/proposal_test.go]:
```suggestion
	// Check that futureTime was not set.
```

File: tests/proposal_test.go [hunk 16]
```diff
@@ -299,29 +394,32 @@ func (s *SwapProposalSuite) BeforeTest(suiteName, testName string) {
...
 	)
 
 	// Check that proposer was set.
 	proposer, err := proposal.Proposer(nil)
 	s.Require().NoError(err)
 	s.Equal(s.proposer.address(), proposer)
 
-	// Check that time was not set.
-	time, err := proposal.Time(nil)
+	// Check that futureTimewas not set.
```
Comment (fiddlemath) [inline tests/proposal_test.go]:
```suggestion
	// Check that futureTime was not set.
```

File: tests/proposal_test.go [hunk 17]
```diff
@@ -349,23 +447,95 @@ func (s *SwapProposalSuite) BeforeTest(suiteName, testName string) {
...
+	futureTime := bigInt(0)
 
 	// Accept proposal.
-	s.requireTxStrongly(s.proposal.Accept(s.signer, time))()
+	s.requireTxWithStrictEvents(s.proposal.Accept(s.signer, futureTime))(
+		abi.SwapProposalProposalAccepted{Proposer: s.proposer.address(), Time: futureTime},
+	)
 
-	// Check that the time was set.
+	// Check that the futureTimewas set.
```
Comment (fiddlemath) [inline tests/proposal_test.go]:
```suggestion
	// Check that the futureTime was set.
```

**Review by fiddlemath — APPROVED** [review]

**Pull request merged**